### PR TITLE
Import current snapshot of QUICHE as an external dependency.

### PR DESCRIPTION
**Pull request by mpwarres**
*Description*:

[QUICHE](https://quiche.googlesource.com/quiche/) is Google's implementation of
QUIC and related protocols. It is the same code used in Chromium and Google's
servers, but packaged in a form that is intended to be easier to incorporate
into third-party projects.

QUICHE is still a work in progress, with current efforts focusing on
eliminating diffs between Google-internal and Chromium variants of the code, as
well as providing buildfiles and a default implementation of the QUICHE
platform API (the platform-dependent porting layer that sits underneath the
main body of QUICHE, which is platform-independent).

In the interim, the Google QUIC team has created a branch of QUICHE,
[envoy-integration](https://quiche.googlesource.com/quiche/+/envoy-integration),
to provide access to the full set of QUICHE source files, while the work above
continues. Since Envoy can overlay its own build rules on top of external
dependencies, this should allow Envoy/QUICHE integration work to move forwards,
without being blocked by completion of "full" QUICHE. When QUICHE is fully
complete, we can just update the dependency to point to the official QUICHE
release, making whatever buildfile adjustments are necessary.

This PR is limited to setting up a basic import of QUICHE with a bare skeleton
of build rules to build and test one element of the platform API. The plan is
to add more build rules and platform implementation classes incrementally, in a
series of follow-on CLs. The end milestone of the platform implementation phase
will be when we can run the QUIC end-to-end tests included in QUICHE on top of
the Envoy platform implementation.

Signed-off-by: Michael Warres <mpw@google.com>

For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

*Risk Level*: low--code not used yet
*Testing*: unit test to exercise build rules
*Docs Changes*: none
*Release Notes*: none
[Optional Fixes #Issue]
[Optional *Deprecated*:]

### CODE REVIEW

**Comment by mpwarres** [comment]
FYI @mattklein123 @alyssawilk @conqerAtapple

**Comment by alyssawilk** [comment]
I think on this one I'd like @htuch to take a pass since it's mostly bazel hackery and I want to make sure all QUIC code is easy to compile out.
Also you'll want to check the builds - looks like format and clang_tidy need a pass.

**Comment by alyssawilk** [comment]
Also builds or not, this PR is *super* exciting.  Hurrah for QUIC integration!

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -191,4 +191,9 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "subpar-1.3.0",
         urls = ["https://github.com/google/subpar/archive/1.3.0.tar.gz"],
     ),
+    com_google_quiche = dict(
```
Comment (htuch) [inline bazel/repository_locations.bzl]: Should this be `com_googlesource_quiche` to reflect the convention?

File: source/extensions/quic_listeners/quiche/BUILD [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+load(
+    "//bazel:envoy_build_system.bzl",
+    "envoy_cc_library",
+    "envoy_package",
+)
+
+envoy_package()
+
+envoy_cc_library(
+    name = "dummy_lib",
```
Comment (htuch) [inline source/extensions/quic_listeners/quiche/BUILD]: Is this just a placeholder that will be removed? If so, can you add comments and a TODO to explain this?

File: source/extensions/quic_listeners/quiche/platform/BUILD [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    "envoy_cc_library",
+    "envoy_package",
+)
+
+envoy_package()
+
+envoy_cc_library(
+    name = "http2_platform_impl_lib",
+    hdrs = [
+        "http2_arraysize_impl.h",
```
Comment (htuch) [inline source/extensions/quic_listeners/quiche/platform/BUILD]: Why do we have this enumeration of headers as well as the one in `quiche.BUILD`?

File: bazel/external/quiche.BUILD [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        "quiche/http2/platform/api/http2_ptr_util.h",
+        "quiche/http2/platform/api/http2_reconstruct_object.h",
+        "quiche/http2/platform/api/http2_string.h",
+        "quiche/http2/platform/api/http2_string_piece.h",
+        "quiche/http2/platform/api/http2_string_utils.h",
+        "quiche/http2/platform/api/http2_test_helpers.h",
+    ],
+    visibility = ["//visibility:public"],
+    deps = [
+        "@envoy//source/extensions/quic_listeners/quiche/platform:http2_platform_impl_lib",
```
Comment (htuch) [inline bazel/external/quiche.BUILD]: I'm a bit confused here. Why are we taking an Envoy-proper library and then exporting it out from the external deps?

File: source/extensions/quic_listeners/quiche/platform/http2_reconstruct_object_impl.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+namespace http2 {
```
Comment (htuch) [inline source/extensions/quic_listeners/quiche/platform/http2_reconstruct_object_impl.h]: So, IIUC, you're essentially providing the QUICHE PAL here. This is a bit unfortunate, since it then needs to sit in its own namespace `http2`, which has nothing to do with the rest of the Envoy code base. Should this live in a separate repo?

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (mpwarres) [reply bazel/repository_locations.bzl]: done

File: source/extensions/quic_listeners/quiche/BUILD (hunk 2 above)
Comment (mpwarres) [reply source/extensions/quic_listeners/quiche/BUILD]: done

File: source/extensions/quic_listeners/quiche/platform/http2_reconstruct_object_impl.h (hunk 5 above)
Comment (mpwarres) [reply source/extensions/quic_listeners/quiche/platform/http2_reconstruct_object_impl.h]: Yes. I guess putting it in a separate repo is one possibility, so long as that does not prevent the files from depending on/referencing Envoy types/libraries. Another possibility would be to add flexibility (e.g. macros?) in QUICHE platform api header files to allow override of the namespace in which the impl types are defined.

Let's chat more about this and the other questions in person--I'll try to find time on your calendar sometime tomorrow (though it looks pretty packed already).

**Comment by mpwarres** [comment]
Update: was almost ready to send this back out for review, but now I've hit a snag in the form of linker errors, that might indicate a problem with this approach for importing. I've pushed my latest commits even though they will cause pre-integration checks to fail, to help provide concrete illustration of the issue.

Details: after resolving other CI issues, I figured I might as well fill in some of the more trivial platform impl definitions. In doing so, I discovered that when the newly-added QUICHE platform impl classes in //source/extensions/quic_listeners/quiche/platform attempt depend on Envoy classes, it causes linker errors, since (IIUC) the build target chain of "@com_googlesource_quiche//:http2_platform" -> "@envoy//source/extensions/quic_listeners/quiche/platform:http2_platform_impl_lib" -> "//source/common/common:assert_lib" generates .o files for various Envoy source files under bazel-out/k8-fastbuild/bin/external/envoy/source/, while the same Envoy classes compiled through normal build target chains (which do not got through "@com_googlesource_quiche//:http2_platform") end up in bazel-out/k8-fastbuild/bin/source/. Then, when the linker tries to link all .o files, it sees multiple definitions.

For example:

```
$ bazel build //test/extensions/quic_listeners/...
[...]
/usr/bin/ld.gold: error: bazel-out/k8-fastbuild/bin/source/common/common/_objs/assert_lib/assert.pic.o: multiple definition of 'Envoy::Assert::ActionRegistrationImpl::debug_assertion_failure_record_action_'
/usr/bin/ld.gold: bazel-out/k8-fastbuild/bin/external/envoy/source/common/common/_objs/assert_lib/assert.pic.o: previous definition here
/usr/bin/ld.gold: error: bazel-out/k8-fastbuild/bin/source/common/common/_objs/assert_lib/assert.pic.o: multiple definition of 'Envoy::Assert::setDebugAssertionFailureRecordAction(std::function<void ()>)'
/usr/bin/ld.gold: bazel-out/k8-fastbuild/bin/external/envoy/source/common/common/_objs/assert_lib/assert.pic.o: previous definition here
/usr/bin/ld.gold: error: bazel-out/k8-fastbuild/bin/source/common/common/_objs/assert_lib/assert.pic.o: multiple definition of 'Envoy::Assert::invokeDebugAssertionFailureRecordAction_ForAssertMacroUseOnly()'
/usr/bin/ld.gold: bazel-out/k8-fastbuild/bin/external/envoy/source/common/common/_objs/assert_lib/assert.pic.o: previous definition here
[...]
```

This (unfortunately) kind of makes sense, given the funky way in which Envoy is depending on QUICHE, which (for platform impl) is depending back on Envoy code.

I need to think more about if there's a way to resolve this. If anyone else has ideas/suggestions, please let me know.  Argh...

**Comment by mpwarres** [comment]
I was able to find a solution to the link issue that allows QUICHE platform impl to depend on Envoy build targets--see bazel-discuss@ [mail thread](https://groups.google.com/d/topic/bazel-discuss/UCJ95z6q3Qg/discussion). @htuch, I think this is ready for another look.

**Comment by mpwarres** [comment]
...and it looks like I spoke too soon. While build_image and clang_tidy now succeed, asan and tsan are still broken, due to some difference in the way they are set up. Looking at that now.

File: bazel/external/quiche.BUILD [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    ),
+    visibility = ["//visibility:private"],
+)
+
+# Note: in dependencies below that reference Envoy build targets in the main
+# repository (particularly for QUICHE platform libs), use '@' instead of
+# '@envoy' as the repository identifier. Otherwise, Bazel generates duplicate
+# object files for the same build target (one under
+# bazel-out/.../bin/external/, and one under bazel-out/.../bin/), eventually
+# resulting in link-time errors.
```
Comment (htuch) [inline bazel/external/quiche.BUILD]: Good find!

File: source/extensions/quic_listeners/quiche/platform/http2_bug_tracker_impl.h [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
```
Comment (htuch) [inline source/extensions/quic_listeners/quiche/platform/http2_bug_tracker_impl.h]: Could you add comments to all the platform impl files just after the NOLINE explaining these are part of the QUICHE platform implementation, these are not consumed directly in Envoy and should not be referenced directly by any Envoy code, etc. This would make the style differences below more palatable. Thanks :)

File: source/extensions/quic_listeners/quiche/platform/http2_test_helpers_impl.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// TODO: implement
+
+#define VERIFY_AND_RETURN_SUCCESS(expression) 0
```
Comment (htuch) [inline source/extensions/quic_listeners/quiche/platform/http2_test_helpers_impl.h]: FWIW, I would love these in OSS gtest..

**Comment by mpwarres** [comment]
tl;dr: the linker issues are due to a Bazel bug that has already been fixed,
but is not yet in an official Bazel release. I've confirmed that (with the
QUICHE platform -> Envoy dependency) both regular and asan/tsan builds succeed
with Bazel built at HEAD and the requisite feature flag flipped. Since the fix
is not yet in the Bazel version used by CI, though, I have backed out the
QUICHE platform -> Envoy dependency in this PR (as we do not need it
immediately), and will add it back later on once the Bazel fix is present in
the version we use.

I've also added the QUICHE platform comments requested, so I think this is
ready for another pass.

Details: for the full story, read this bazel-discuss@ [mail
thread](https://groups.google.com/d/topic/bazel-discuss/UCJ95z6q3Qg/discussion). It
turns out we hit a known Bazel issue, bazelbuild/bazel#6848, which (in our
case) caused references made by an external dependency (QUICHE) to
"@envoy//foo:bar" to be considered external, even though "@envoy" designates
the main workspace.

My earlier fix of changing the dependency to use "@//foo:bar" instead of
"@envoy//foo:bar" worked for the default build since it forced the dependency
to refer to the main workspace (i.e. "@"). However, this caused failures in the
CI ASAN and TSAN builds because those build Envoy as an [external
dependency](https://github.com/envoyproxy/envoy-filter-example/blob/6c0a95509091a14981f84f07fa650c3cf3\
2f724e/WORKSPACE#L4)
of [envoy-filter-example](https://github.com/envoyproxy/envoy-filter-example),
so in those setups, "@" resolved to the main workspace of envoy-filter-example,
and not @envoy. (The context-dependent nature of "@" and its implications are
explained most clearly in this
[message](https://groups.google.com/d/msg/bazel-discuss/UCJ95z6q3Qg/dog1mjzlCAAJ)
from the bazel-discuss@ thread.)

Eventually, once the fix for bazelbuild/bazel#6848 is in a release and made
default behavior, dependency references to "@envoy//foo:bar" will resolve to
the right place in both situations.

File: bazel/external/quiche.BUILD (hunk 6 above)
Comment (mpwarres) [reply bazel/external/quiche.BUILD]: See latest top-level comment for an explanation of what's up with TSAN/ASAN. Basically it's because they build Envoy as an external dependency of the filter example.

File: source/extensions/quic_listeners/quiche/platform/http2_bug_tracker_impl.h (hunk 7 above)
Comment (mpwarres) [reply source/extensions/quic_listeners/quiche/platform/http2_bug_tracker_impl.h]: Done.
